src/helpers/util.ts

```diff
@@ -104,17 +104,16 @@ export namespace objectUtil {
   //   [k in keyof T]: undefined extends T[k] ? k : never;
   // }[keyof T];
 
-  type requiredKeys<T extends object> = {
-    [k in keyof T]: undefined extends T[k] ? never : k;
-  }[keyof T];
+  type keepRequiredKeys<T extends object> = {
+    [k in keyof T as undefined extends T[k] ? never : k]-?: T[k];
```

Comment:
if some semantic difference is found using this approach then we could also try this
```suggestion
    [k in keyof T as undefined extends T[k] ? never : k]-?: unknown;
```